models.py

```diff
@@ -280,19 +280,38 @@ def __get_distance_weights(d, w1, w0, sigma):
     return weights
 
 
-def __get_size_weights(sizes):
-    C = Variable(torch.sqrt(torch.Tensor([sizes.size()[1] * sizes.size()[2]])) / 2, requires_grad=False)
-    if torch.cuda.is_available():
-        C = C.cuda()
+def __get_size_weights(sizes, C):
     size_weights = C / sizes
     size_weights[sizes == 1] = 1
     return size_weights
 
 
-def get_loss_params(w0, sigma):
+def get_loss_params(w0, sigma, imsize):
     w0 = Variable(torch.Tensor([w0]), requires_grad=False)
     sigma = Variable(torch.Tensor([sigma]), requires_grad=False)
+    C = Variable(torch.sqrt(torch.Tensor([imsize[0] * imsize[1]])) / 2, requires_grad=False)
     if torch.cuda.is_available():
         w0 = w0.cuda()
         sigma = sigma.cuda()
-    return {'w0': w0, 'sigma': sigma}
+        C = C.cuda()
+    return {'w0': w0, 'sigma': sigma, 'C': C}
+
+
+def mixed_dice_cross_entropy_loss(output, target, dice_weight, ce_weight, ce_loss=None):
+    dice_target = target[:, 0, :, :].long()
+    ce_target = target
+    if ce_loss is None:
+        ce_loss = torch.nn.CrossEntropyLoss()
+        ce_target = dice_target
+    return dice_weight * dice_loss(output, dice_target) + ce_weight * ce_loss(output, ce_target)
+
+
+def dice_loss(output, target):
+    dice_numerator = 0
+    dice_denominator = 0
+    for class_nr in range(1, int(target.max()) + 1):
```

Comment:
@apyskir this assumes that 0 is background. Could case trouble when we encode/preprocess stuff differently

Reply:
@jakubczakon Yes, I think that this is a natural and justified assumption. Why would you mark background with different label? In case of binary models (like in DSB) you also made this assumption.